File: src/main/java/io/mazenmc/skypebot/stat/MessageStatistic.java
package io.mazenmc.skypebot.stat;

import com.skype.ChatMessage;
import com.skype.SkypeException;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class MessageStatistic {
    private final String name;
    private final List<Message> messages;

    MessageStatistic(String name) {
        this.name = name;
        messages = new ArrayList<>();
    }

    MessageStatistic(String name, JSONArray messages) {
        this(name);

        IntStream.range(0, messages.length()).forEach((i) -> {
            try {
                JSONObject object = messages.getJSONObject(i);

                this.messages.add(new Message(object.getString("contents"), object.getLong("time")));
            } catch (JSONException ex) {
                ex.printStackTrace();
            }
        });

        if (this.messages.isEmpty()) {
            this.messages.add(new Message("<never sent message>", System.currentTimeMillis()));
        }
    }

    public String name() {
        return name;
    }

    public List<String> words() {
        List<String> words = new ArrayList<>();

        messages.stream()
                .map((s) -> s.contents().split(" "))
                .filter((s) -> !s.trim().isEmpty())
                .forEach((s) -> words.addAll(Arrays.asList(s)));

        return words;
    }

    public String mostCommonWord() {
        return words().stream()
                .collect(Collectors.groupingBy(w -> w, Collectors.counting()))
                .entrySet().stream()
                .sorted((e, e1) -> (int) (e1.getValue() - e.getValue()))
                .findFirst().get()
                .getKey();
    }

    public int wordCount() {
        return messages.stream()
                .mapToInt((m) -> m.contents().split(" ").length)
                .filter(i -> i < 50 && i > 2)
                .sum();
    }

    public double averageWords() {
        return wordCount() / messageAmount();
    }

    public int messageAmount() {
        return messages.size();
    }

    public Message randomMessage() {
        return messages.get(ThreadLocalRandom.current().nextInt(messages.size()));
    }

    public int commandCount() {
        return (int) messages.stream()
                .map(Message::contents)
                .filter((s) -> s.startsWith("@"))
                .count();
    }

    public double commandPercent() {
        return ((double) commandCount() / (double) messageAmount()) * 100;
    }

    public void addMessage(ChatMessage message) throws SkypeException {
        messages.add(new Message(message.getContent(), message.getTime().getTime()));
    }

    public void addMessage(Message message) {
        messages.add(message);
    }

    public List<Message> messages() {
        return messages;
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof MessageStatistic && messages.equals(((MessageStatistic) obj).messages);
    }
}

File: src/main/java/io/mazenmc/skypebot/stat/MessageStatistic.kt
package io.mazenmc.skypebot.stat

import com.samczsun.skype4j.chat.messages.ReceivedMessage
import org.json.JSONArray
import java.util.*
import java.util.concurrent.ThreadLocalRandom
import java.util.stream.Collectors
import java.util.stream.IntStream

public class MessageStatistic(var name: String) {
    private var messages: MutableList<Message> = ArrayList<Message>()

    constructor(name: String, messages: JSONArray) : this(name) {
        IntStream.range(0, messages.length()).forEach { i ->
            var obj = messages.getJSONObject(i)
            var contents = obj.getString("contents")

            if (!contents.equals("<never sent message>")) {
                this.messages.add(Message(contents, obj.getLong("time")))
            }
        }

        if (this.messages.isEmpty()) {
            this.messages.add(Message("<never sent message>", System.currentTimeMillis()))
        }
    }

    public fun name(): String {
        return name
    }

    public fun words(): List<String> {
        var words = ArrayList<String>()

        messages.map { s -> s.contents.splitBy(" ") }.forEach { s -> s.forEach { w -> words.add(w) } }

        return words
    }

    public fun wordCount(): Int {
        return messages.map { m -> m.contents().splitBy(" ").size() }
                .filter { i -> i > 2 }
                .sum()
    }
    
    public fun letterCount(): Int {
        return messages.map { m -> m.contents().toCharList().filter { i -> i != ' ' }.size() }
                .sum()
    }
    
    public fun messageAmount(): Int {
        return messages.size()
    }
    
    public fun averageWords(): Int {
        return wordCount() / messageAmount()
    }
    
    public fun randomMessage(): Message {
        return messages.get(ThreadLocalRandom.current().nextInt(messageAmount()))
    }
    
    public fun commandCount(): Int {
        return messages.map(Message::contents)
                .filter { s -> s.startsWith("@") }
                .size()
    }

    public fun commandPercent(): Double {
        return (commandCount().toDouble() / messageAmount().toDouble()) * 100
    }

    public fun addMessage(message: ReceivedMessage) {
        addMessage(Message(message.getContent().asPlaintext(), System.currentTimeMillis()))
    }

    public fun addMessage(message: Message) {
        messages.add(message)
    }

    public fun messages(): List<Message> {
        return messages
    }
}